src/middleware/company.authentication.middleware.ts

```diff
@@ -0,0 +1,32 @@
+import { Request, Response, NextFunction } from 'express';
+import { logger } from '../utils/logger';
+import { authMiddleware, AuthOptions } from "@companieshouse/web-security-node";
+import { CHS_URL, OVERSEAS_ENTITY_PRESENTER_URL } from '../config';
+import { getApplicationData } from "../utils/application.data";
+import { ApplicationData } from "../model";
+import { EntityNumberKey } from "../model/data.types.model";
+
+export const companyAuthentication = (req: Request, res: Response, next: NextFunction) => {
+  try {
+    logger.debugRequest(req, `COMPANY AUTHENTICATION REQUEST`);
+
+    const appData: ApplicationData = getApplicationData(req.session);
+    const entityNumber: string = appData?.[EntityNumberKey] as string;
+
+    if (entityNumber) {
+      const authMiddlewareConfig: AuthOptions = {
+        chsWebUrl: CHS_URL,
+        returnUrl: OVERSEAS_ENTITY_PRESENTER_URL,
+        companyNumber: entityNumber
+      };
+      logger.infoRequest(req, `Invoking company authentication with (${ entityNumber }) present in session`);
+      return authMiddleware(authMiddlewareConfig)(req, res, next);
```

Comment:
By using the `web-security-node` you will end-up having duplicating checks between our authentication middleware and the new one.
I would have created locally a separate middleware with a simple logic, similar to [here](https://github.com/companieshouse/web-security-node/blob/5acbe195eea859a0021e2d911f35637f2f533bb7/src/index.ts#L59)

```
function isAuthorisedForCompany(companyNumber: string, signInInfo: ISignInInfo): boolean {
  const authorisedCompany = signInInfo[SignInInfoKeys.CompanyNumber]
  if (!authorisedCompany) {
    return false
  }

  return authorisedCompany.localeCompare(companyNumber) === 0
}
```

Reply:
@mouhajer-ch ,  but won't this expose the checking logic to overseas-web which currently delegates the checking logic to web-security-node

Reply:
What do you mean?

Reply:
@mouhajer-ch This logic currently resides in web-security-node , so was not sure if we need to add this into overseas-web

Reply:
Nope

Reply:
FYI we are using the current pattern used in other CH AuthCode integrations such as confirmation-statement

Reply:
@jkayyala I think @mouhajer-ch suggestion on review is a good one --
we also do the authenticated check in our middleware and REDIRECT only if not authenticated. This does save trips/redirects to just do the authenticated checks !

    if (entityNumber && !isAuthorisedForCompany ){
      const authMiddlewareConfig: AuthOptions = {
        chsWebUrl: CHS_URL,
        returnUrl: OVERSEAS_ENTITY_PRESENTER_URL,

Reply:
I agree with @mouhajer-ch that the current implementation within the `web-security-node` `authMiddleware` does not seem ideal, as it bundles user _authentication_ checks together with company _authorisation_ checks (and we are already handling user authentication in our app).

Appreciate that you are just following the same pattern as used in other CH web apps but I think it would worth be looking at the solution that implements a local company auth middleware check. If it doesn't work, fine, we can fallback to this approach but worthy of consideration I think.

Reply:
@mwestacott , implemented the changes @mouhajer-ch had suggested, can you please review

Reply:
Resolving this comment as covered by newer comments below.